src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala

```diff
@@ -112,44 +153,331 @@ class AppsController(
+        } else {
+          onSuccess(taskKiller.kill(appId, findToKill, wipe)) { instances =>
+            complete(Json.obj("tasks" -> instances))
+          }
+        }
+    }
+  }
+
+  private def killTask(appId: PathId, taskId: TaskId)(implicit identity: Identity) = {
+    //the line below doesn't look nice but it doesn't compile if we use parameters directive
+    (forceParameter & parameter("host") & parameter("scale".as[Boolean].?(false)) & parameter("wipe".as[Boolean].?(false))) {
+      (force, host, scale, wipe) =>
+
+        def findToKill(appTasks: Seq[Instance]): Seq[Instance] = {
+          try {
+            val instanceId = taskId.instanceId
+            appTasks.filter(_.instanceId == instanceId)
+          } catch {
+            // the id can not be translated to an instanceId
+            case _: MatchError => Seq.empty
+          }
+        }
+
+        if (scale && wipe) throw new BadRequestException("You cannot use scale and wipe at the same time.")
+
+        if (scale) {
+          val deploymentPlanF = taskKiller.killAndScale(appId, findToKill, force)
+          onSuccess(deploymentPlanF) { plan =>
+            complete((StatusCodes.OK, List(Headers.`Marathon-Deployment-Id`(plan.id)), Messages.DeploymentResult(plan)))
+          }
+        } else {
+          onSuccess(taskKiller.kill(appId, findToKill, wipe)) { instances =>
+            complete(Json.obj("tasks" -> instances))
+          }
+        }
+    }
+  }
+
+  private def listVersions(appId: PathId)(implicit identity: Identity) = {
+    val versions = groupManager.appVersions(appId).runWith(Sink.seq)
+    authorized(ViewRunSpec, groupManager.app(appId), Rejections.EntityNotFound.app(appId)).apply {
+      onSuccess(versions) { versions =>
+        complete(Json.obj("versions" -> versions))
+      }
+    }
+  }
+
+  private def getVersion(appId: PathId, version: Timestamp)(implicit identity: Identity) = {
+    onSuccess(groupManager.appVersion(appId, version.toOffsetDateTime)) {
+      case Some(app) =>
+        authorized(ViewRunSpec, app, Rejections.EntityNotFound.app(appId)).apply {
+          complete(app)
+        }
+      case None =>
+        reject(Rejections.EntityNotFound.app(appId))
+    }
+  }
 
+  // format: OFF
   val route: Route = {
     asLeader(electionService) {
       authenticated.apply { implicit identity =>
-        pathEnd {
-          post {
-            (entity(as[AppDefinition]) & parameters('force.as[Boolean].?(false))) { (app, force) =>
-              createApp(app, force)
-            }
+        pathEndOrSingleSlash {
+          get {
+            listApps
           } ~
-            get {
-              listApps
-            }
+          put {
+            replaceMultipleApps
+          } ~
+          patch {
+            patchMultipleApps
+          } ~
+          post {
+            createApp
+          }
         } ~
-          path(RemainingPathId) { appId =>
+        extractExistingAppId(groupManager.rootGroup()) { appId =>
```

Comment:
not very familiar with akkahttp but the `rootGroup` is fetched every time, right?

Reply:
It seems so